src/main/kotlin/fp/Maybe/Maybe.kt

```diff
@@ -1,55 +1,31 @@
 package fp.Maybe
 
-sealed abstract class Maybe<T> {
-  companion object {
-    fun <T> of(value : T) : Just<T> = Just(value)
-
-    fun <T> nothing() : Nothing<T> = Nothing()
+sealed class Maybe<out A> {
 
-    fun <T> fromNullable(value : T?) : Maybe<T> =
-      if (value != null) Just(value) else Nothing()
+  companion object {
+    inline fun <reified A : Any> fromNullable(a: A?) : Maybe<A> = if (a != null) Maybe.Just(a) else Maybe.None
+    operator fun <A> invoke(a: A): Maybe<A> = Maybe.Just(a)
   }
 
-  abstract fun isJust() : Boolean
-
-  abstract fun isNothing() : Boolean
-
-  abstract fun getValue() : T?
-
-  abstract fun map(fn : (T) -> T?) : Maybe<T>
-
-  abstract fun apply(fn : (T) -> Any?) : Maybe<Any?>
-
-  abstract fun getOrElse(other : Any? = null) : Any?
-
-}
-
-class Just<T>(private val value : T) : Maybe<T>() {
-  override fun isJust() : Boolean = true
+  abstract val isEmpty : Boolean
 
-  override fun isNothing() : Boolean = false
+  inline fun <B> fold(ifEmpty : () -> B, fn: (A) -> B): B = when (this) {
+    is Maybe.None -> ifEmpty()
+    is Maybe.Just -> fn(value)
+  }
 
-  override fun getValue(): T? = value
+  inline fun <B> map(fn : (A) -> B) : Maybe<B> = fold({ Maybe.None }, { a -> Maybe.Just(fn(a)) })
 
-  override fun map(fn: (T) -> T?): Maybe<T> = fromNullable(fn(value))
+  inline fun <B> flatMap(fn : (A) -> Maybe<B>): Maybe<B> = fold({ Maybe.None }, { a -> fn(a) })
 
-  override fun apply(fn: (T) -> Any?): Maybe<Any?> = fromNullable(fn(value))
+  data class Just<out A>(val value: A) : Maybe<A>() {
+    override val isEmpty: Boolean = false
+  }
 
-  override fun getOrElse(other: Any?): Any? = value as Any
+  object None : Maybe<Nothing>() {
```

Comment:
Nothing is already a primitive type in Kotlin and I was having some issues because of that.

You are correct, this monad is called Maybe in Haskell.